packages/blade/src/components/Tour/utils.ts

```diff
@@ -0,0 +1,148 @@
+    }, delay);
+
+    return () => {
+      window.clearTimeout(timeoutRef.current);
+    };
+  }, [delay, initialState]);
+
+  return delayedState;
+}
+
+/**
+ * Keep track of when we are transitioning between steps
+ *
+ * This is used to prevent the popover from jumping to the next step before animations are finished
+ */
+const useIsTransitioningBetweenSteps = (activeStep: number, transitionDelay: number) => {
+  const prevActiveStep = usePrevious(activeStep);
+  const [isTransitioning, setIsTransitioning] = React.useState(false);
+
+  // Keep track of when we are transitioning between steps
+  React.useEffect(() => {
+    if (prevActiveStep === undefined) return;
+    setIsTransitioning(true);
+    const timeout = setTimeout(() => {
+      setIsTransitioning(false);
+    }, transitionDelay);
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [prevActiveStep, transitionDelay]);
+
+  return isTransitioning;
+};
+
+// https://stackoverflow.com/questions/46795955/how-to-know-scroll-to-element-is-done-in-javascript
+function smoothScroll(element: Element | null, options: ScrollIntoViewOptions) {
+  return new Promise((resolve) => {
+    if (__DEV__) {
+      if (!(element instanceof Element)) {
+        throwBladeError({
+          moduleName: 'smoothScroll',
+          message: 'argument "element" must be an instance of Element',
+        });
+      }
+    }
+
+    let same = 0;
+    let lastPos: undefined | null | number = null;
+    const scrollOptions = { behavior: 'smooth', ...options } as const;
+
+    element!.scrollIntoView(scrollOptions);
+    requestAnimationFrame(check);
+
+    // eslint-disable-next-line consistent-return
+    function check() {
+      const newPos = element?.getBoundingClientRect().top;
+      if (newPos === lastPos) {
+        if (same++ > 2) {
+          return resolve(null);
+        }
+      } else {
+        same = 0;
+        lastPos = newPos;
+      }
+      requestAnimationFrame(check);
+    }
+  });
+}
+
+function useIntersectionObserver(
+  elementRef: React.RefObject<Element>,
+  { threshold = 0, root = null, rootMargin = '0%' }: IntersectionObserverInit,
+): IntersectionObserverEntry | undefined {
+  const [entry, setEntry] = React.useState<IntersectionObserverEntry>();
+
+  const updateEntry = ([entry]: IntersectionObserverEntry[]): void => {
+    setEntry(entry);
+  };
+
+  React.useEffect(() => {
+    const node = elementRef?.current; // DOM Ref
+    const hasIOSupport = !!window.IntersectionObserver;
+
+    if (!hasIOSupport || !node) return;
+
+    const observerParams = { threshold, root, rootMargin };
+    const observer = new IntersectionObserver(updateEntry, observerParams);
+
+    observer.observe(node);
+
+    return () => observer.disconnect();
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [elementRef?.current, JSON.stringify(threshold), root, rootMargin]);
+
+  return entry;
+}
+
+const useLockBodyScroll = (shouldLock: boolean) => {
```

Comment:
Does this have proper browser support? Will the logic break in any older browsers?

Reply:
It's not a browser API, it's a library we are using. `body-scroll-lock` 

This is already being used in BottomSheet too.

Reply:
why do we need to create this as a util on top of another util that we only have created? something special we want to do?

Reply:
`useScrollLock` is more generic and gives us the manual control via ref to enable or disable the body scroll lock. 

I created a light wrapper around it to provide a more cleaner API which just enables/disables the lock depending on the isOpen prop of Tour. 

I can inline this, but i don't want this random useEffect to be on the body of the already complicated to look at Tour component, reusing this makes Tour.web.tsx code easier to reason about and cleaner.